AppCenter/AppCenter/Internals/Storage/MSLogDBStorage.m

```diff
@@ -67,23 +67,29 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupId:(NSString *)groupId flags:(MSFlags)fl
                                    targetKey ? [NSString stringWithFormat:@"'%@'", targetKey] : @"NULL", (unsigned int)persistenceFlags];
   }
   return [self executeQueryUsingBlock:^int(void *db) {
+           // Selecting logs with equal or lower priority and ordering by priority then age.
            int result = [MSDBStorage executeNonSelectionQuery:addLogQuery inOpenedDatabase:db];
 
            // If the database is full, delete logs until there is room to add the log.
            long countOfLogsDeleted = 0;
-           while (result == SQLITE_FULL) {
-             result = [MSLogDBStorage deleteOldestLogsWithCount:1 inOpenedDatabase:db];
+           NSArray<NSNumber *> *ids = [self logsWithPriorityLowerThan:persistenceFlags inOpenedDatabase:db];
+           NSUInteger index = 0;
+           while (result == SQLITE_FULL && index < [ids count]) {
+             result = [MSLogDBStorage deleteLogsFromDBWithColumnValues:@[ ids[index++] ] columnName:kMSIdColumnName inOpenedDatabase:db];
              if (result != SQLITE_OK) {
                break;
              }
              ++countOfLogsDeleted;
              result = [MSDBStorage executeNonSelectionQuery:addLogQuery inOpenedDatabase:db];
            }
            if (countOfLogsDeleted > 0) {
-             MSLogDebug([MSAppCenter logTag], @"Log storage was over capacity, %ld oldest log(s) deleted.", (long)countOfLogsDeleted);
+             MSLogDebug([MSAppCenter logTag], @"Log storage was over capacity, %ld oldest log(s) with equal or lower priority deleted.",
+                        (long)countOfLogsDeleted);
            }
            if (result == SQLITE_OK) {
              MSLogVerbose([MSAppCenter logTag], @"Log is stored with id: '%ld'", (long)sqlite3_last_insert_rowid(db));
+           } else if (result == SQLITE_FULL && index == [ids count]) {
+             MSLogDebug([MSAppCenter logTag], @"No logs with equal or lower priority found and storage already full; discarding log.");
```

Comment:
Just reading through the code but `result == SQLITE_FULL && index == [ids count]` can happen for two reasons. First as described "No logs with equal or lower priority found and storage already full", that's when `index` and `[ids count]` are both 0. But the second situation is when `index` and `[ids count]` are equals but > 0, it happens when all the lower priority logs got deleted but that wasn't enough and the DB is still full. So maybe we want to revisit the logging here to log those two situations separately.

Reply:
Both conditions are actually same. Eventually the log will be discarded and the only difference is that there are some logs deleted before discarding the log.